jax/custom_derivatives.py

```diff
@@ -951,3 +955,162 @@ def merge(l1, l2):
 
 def abstractify(x):
   return core.raise_to_shaped(core.get_aval(x))
+
+
+### Custom transposition
+
+def linear_call(fun: Callable, fun_transpose: Callable, residual_args,
+                linear_args):
+  """Call a function, indicating a custom implementation for its transpose.
+
+  The type signatures of ``fun`` and ``fun_transpose`` are:
+
+  .. code-block:: haskell
+
+    fun           :: r -> a -o b
+    fun_transpose :: r -> b -o a
+
+  where the ``-o`` arrow indicates a linear function, ``r`` is the
+  residual input type and ``a`` is the linear input type.
+
+  The functions ``fun`` and ``fun_transpose`` are coupled as
+  transposes of one another across transposition. Specifically, the
+  transpose of a ``linear_call`` primitive is another ``linear_call``
+  to ``fun_transpose``, with ``fun`` as its custom transposition.
+
+  For example, if::
+
+    def f(r, x):
+      return x / r
+
+    def t(r, t):
+      return t / r
+
+    def div_add(denom, x):
+      return x + linear_call(f, t, denom, x)
+
+    def transpose(f, x_example):
+      def transposed(y):
+        x, = jax.linear_transpose(f, x_example)(y)
+        return x
+      return transposed
+
+  Then:
+
+  >>> div_add(9., 3.)
+  12.0
+  >>> transpose(partial(div_add, 3.), 1.)(18.)  # custom
+  24.0
+  >>> transpose(lambda x: x + x / 3., 1.)(18.)  # reference
+  24.0
+
+  As another example, if::
+
+    def custom_id(x):
+      def f(_, x): return x
+      def t(_, t): return 7.
+      return linear_call(f, t, (), x)
+
+  Then:
+
+  >>> custom_id(1.)
+  1.
+  >>> transpose(custom_id, 1.)(1.)
+  7.
+  >>> transpose(transpose(custom_id, 1.), 1.)(1.)
+  1.
+  >>> transpose(transpose(transpose(custom_id, 1.), 1.), 1.)(1.)
+  7.
+
+  Args:
+    fun: a Python callable specifying a structurally linear
+      function. It should take two arguments: one of "residual" inputs
+      (type ``r``), i.e. inputs in which the function is not
+      necessarly linear, and one of "linear" inputs (type ``a``).  It
+      should return output whose components are linear in the linear
+      input (type ``b``).
+    fun_transpose: a Python callable specifying a structurally linear
+      function that is the transpose of ``fun`` with respect to its
+      linear inputs. Its first argument is the same residual inputs
+      (``r``) as ``fun``. Its second argument is of type
+      ``b``. Finally, its output is of type ``a`` and each of its
+      component are linear in its second argument (the ``b`` inputs).
```

Comment:
Document `residual_args` and `linear_args`? The API may change but we'll probably have to document them somewhere, so might be good to cut a draft now. Also, I suspect that "residual" might be one of the things most in need of documentation.